src/Native/System.Native/pal_readdir.cpp

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "../config.h"
+#include "pal_readdir.h"
+#include <errno.h>
+
+static void ConvertDirent(const dirent& entry, DirectoryEntry* outputEntry)
+{
+    // We use Marshal.PtrToStringAnsi on the managed side, which takes a pointer to
+    // the start of the unmanaged string. Give the caller back a pointer to the 
+    // location of the start of the string that exists in their own byte buffer.
+    outputEntry->nameStartAddress = &entry.d_name[0];
+    outputEntry->inode_type = (NodeType)entry.d_type;
+
+#if HAVE_DIRENT_NAME_LEN
+    outputEntry->nameLength = entry.d_namlen;
+#else
+    outputEntry->nameLength = -1; // sentinel value to mean we have to walk to find the first \0
+#endif
+}
+
+extern "C" 
+int32_t GetDirentSize()
+{
+    // Truncating this should be fine
```

Comment:
static_assert(sizeof(dirent) <= INT32_MAX, "")? We need it to be much smaller than this of (we stack alloc it) and code samples everywhere do likewise so it's surely small enough. Just prefer asserts of "should be fine comment".